lib/teleterm/api/uri/uri.go

```diff
@@ -189,6 +189,14 @@ func (r ResourceURI) AppendKube(name string) ResourceURI {
 	return r
 }
 
+// AppendKubeResource appends kube resource segment to the URI.
+// Modeled after how access request constructs the resource ID:
+// <teleport-cluster-name>/<subresource kind>/<kube_cluster name>/<subresource name>
+func (r ResourceURI) AppendKubeResource(kubeResourceKind string, kubeClusterName string, kubeResourceName string) ResourceURI {
+	r.path = fmt.Sprintf("%v/%v/%v/%v", r.path, kubeResourceKind, kubeClusterName, kubeResourceName)
+	return r
+}
+
```

Comment:
i wasn't entirely sure what this was for, other than perhaps to uniquely identify resources?

Reply:
Yeah, URIs are used for identifying resources.
```
fmt.Sprintf("%v/%v/%v/%v", r.path, kubeResourceKind, kubeClusterName, kubeResourceName)
```
I see some problems with this scheme. `kubeResourceKind` is an unknown string, so in theory, it could collide with other URIs. Additionally, we would not be able to identify if a given URI string is a kube resource URI without knowing `kubeResourceKind`.
A fix for this would be adding a known prefix in the URI, e.g:
```
"/clusters/teleport.sh/kube-resources/namespace/kube-cluster-name/namespace-name"
```

However, perhaps it would make more sense to treat a kube resource as a child of the kube cluster? I mean something like this:
```go
var pathKubes = urlpath.New("/clusters/:cluster/kubes/:kubeName")
var pathKubeResources = urlpath.New("/clusters/:cluster/kubes/:kubeName/:subResourceKind/:subResourceName")
```

cc @ravicious

Reply:
i went ahead and applied your first suggestion instead:

```
"/clusters/teleport.sh/kube-resources/namespace/kube-cluster-name/namespace-name"
```

Reply:
If it's correct that kube subresources always belong to a specific kube resource, and I believe it's correct, then I'd definitely go with the second option that Grzegorz proposed.

If we currently support only namespaces within the app, then for now I'd make a definition only for namespaces. That would be:

```typescript
  kubeResourceNamespace:
    '/clusters/:rootClusterId/(leaves)?/:leafClusterId?/kubes/:kubeId/namespaces/:kubeNamespaceId',
  kubeResourceNamespaceLeaf:
    '/clusters/:rootClusterId/leaves/:leafClusterId/kubes/:kubeId/namespaces/:kubeNameSpaceId',
```

This follows the existing conventions that we have while introducing a new one. Nested identifiers are prefixed with "parent's" name, e.g. it's `:kubeNamespaceId`, not `:namespaceId`.

This is mostly due to how `matchPath` works and how we implemented `parseUri`. You give it a bunch of possible identifiers and then pluck out what you need. So if we end up having two URIs, I think we'd rather have two different fields prefixed with the parent name, rather than reusing the same field name for two different purposes.

I wouldn't bother with `KubeResourceNamespaceParam` for now. If we wanted to have better types around this, we should refactor how `routing.parseUri` works and all the helper functions like `routing.parseServerUri`.

---

On the Go side, `AppendKubeResourceNamespace` as you have it is fine, we'd only have to change the actual URI it generates.